_posts/2025-05-08-agentic-ai-with-quarkus-p3.adoc

```diff
@@ -0,0 +1,115 @@
+
+In essence, the most relevant difference between the two is that workflow patterns are programmatically orchestrated through handcrafted code paths, while agents autonomously decide their own processes and tool usage, maintaining control over how they execute tasks. This makes them more flexible and adaptable to various scenarios, but it also makes them less predictable and, in some cases, more prone to hallucinations.
+
+[.text-center]
+.AI Workflow Vs. Pure Agentic AI
+image::wrokflowVsAgents.png[align="center", alt="AI Workflow Vs. Pure Agentic AI"]
+
+== From Workflow to Agents
+
+https://quarkus.io/blog/agentic-ai-with-quarkus/#routing[Routing] is one of the workflow patterns presented in the first article of this series. There, we used a first LLM service to categorize the user request and then used that category to programmatically reroute that request to one of three other LLMs acting as medical, legal, or technical experts.
+
+[.text-center]
+.Routing pattern
+image::routing-pattern.png[width=70%, align="center", alt="Routing pattern"]
+
+In that example, each expert was implemented as a separate and independent LLM service, and the routing to one of them was performed programmatically by the application code. Tracing the execution of a request like:
+
+[source, shell]
+----
+curl http://localhost:8080/expert/request/I%20broke%20my%20leg%20what%20should%20I%20do
+----
+
+The resulting traces show the sequence of steps performed to fulfill the user request: first, the relatively fast, less than 2 seconds, classification phase conducted by the Router Agent, then the more expensive invocation of the selected expert service, which took almost 25 seconds to generate its answer.
+
+[.text-center]
+.Tracing routing workflow pattern execution
+image::routing-workflow-trace.png[align=center, alt="Tracing routing workflow pattern execution"]
+In this example, we use the same model to classify and generate the response. However, it is possible to use specialized models for the classification and each experts.
+Now, let’s see how we can transform this workflow approach into a more agentic one.
+
+Quarkus integration makes it straightforward to turn these AI “expert” services into _tools_ that another AI service can invoke. You only need to annotate the AI service methods with `@Tool` and configure the caller AI service with `@Toolbox`. This approach preserves the possibility of also invoking the single expert directly as an independent LLM service, and also using specialized models for each expert.
+
+[source, java]
+----
+public interface MedicalExpert {
+
+    @UserMessage("""
+            You are a medical expert.
+            Analyze the following user request under a medical point of view and provide the best possible answer.
+            The user request is {request}.
+            """)
+    @Tool("A medical expert") // <-- Allows to use this LLM also as a tool for other LLMs
```

Comment:
Is this the MCP `@Tool` annotation?

Reply:
Okay, maybe just add a note that mention that this is regular tool like any other and not an MCP tool (which as you said will be described in a later blog post).
The reason here is that many people have hearf of MCP these days so they will probably be wondering